Main branch failing `deploy` task due to mvn error

It looks like we have not had a successful run of the `deploy` task for this repo in about a week, failing with a maven error while building the docker image. Strangely, the `deploy-to-private-gcr` task (which also has to build the image) works.

I notice that the docker build is using cache up until the failing `RUN mvn dependency:copy-dependencies` invocation, so perhaps there's something bad in the docker layer cache?

The error is:

```
Step 12/23 : RUN mvn dependency:copy-dependencies
 ---> Running in 2efdc878229b
ls: cannot access '/usr/bin/mvn': Operation not permitted
Error: Could not find or load main class org.codehaus.plexus.classworlds.launcher.Launcher
Caused by: java.lang.ClassNotFoundException: org.codehaus.plexus.classworlds.launcher.Launcher
The command '/bin/sh -c mvn dependency:copy-dependencies' returned a non-zero code: 1
```

cc @relud

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -3,6 +3,7 @@ version: 2.1
 
 orbs:
   gcp-gcr: circleci/gcp-gcr@0.13.0
+  docker: circleci/docker@1.5
 
 jobs:
   build:
@@ -317,9 +318,7 @@ jobs:
               && git push \
               || echo "Skipping push since it looks like there were no changes"
   deploy:
-    docker:
-      # bash required for step: Determine docker image name
-      - image: relud/docker-bash
+    executor: docker/docker
     steps:
       - checkout
       - setup_remote_docker:
@@ -331,20 +330,22 @@ jobs:
           command: |
             rm -rf sql/
             cp -r /tmp/workspace/generated-sql/sql sql
-      - run:
-          name: Determine docker image name
-          command:
-            # yamllint disable-line rule:line-length
-            echo 'IMAGE="${CIRCLE_PROJECT_USERNAME+$CIRCLE_PROJECT_USERNAME/}${CIRCLE_PROJECT_REPONAME:-bigquery-etl}:${CIRCLE_TAG:-latest}"' > $BASH_ENV
-      - run:
-          name: Build docker image
-          command: docker build . --pull --tag "$IMAGE"
-      - run:
-          name: Deploy to Dockerhub
-          command: |
-           echo "${DOCKER_PASS:?}" | \
-            docker login -u "${DOCKER_USER:?}" --password-stdin
-            docker push "$IMAGE"
+      - when:
+          condition: &deploy-condition
+            or:
+              - equal: [main, << pipeline.git.branch >>]
+              - << pipeline.git.tag >>
+          steps:
+            - docker/check:
+                docker-password: DOCKER_PASS
+                docker-username: DOCKER_USER
+      - docker/build: &public-image
+          image: ${CIRCLE_PROJECT_USERNAME+$CIRCLE_PROJECT_USERNAME/}${CIRCLE_PROJECT_REPONAME:-bigquery-etl}
+          tag: ${CIRCLE_TAG:-latest}
+      - when:
+          condition: *deploy-condition
+          steps:
+            - docker/push: *public-image
   private-generate-sql:
     docker: *docker
     steps:
@@ -419,13 +420,17 @@ jobs:
           command: |
             rm -rf sql/
             cp -r /tmp/workspace/private-generated-sql/sql sql
-      - gcp-gcr/gcr-auth
-      - gcp-gcr/build-image:
+      - when:
+          condition: *deploy-condition
+          steps:
+            - gcp-gcr/gcr-auth
+      - gcp-gcr/build-image: &private-image
           image: bigquery-etl
           tag: ${CIRCLE_TAG:-latest}
-      - gcp-gcr/push-image:
-          image: bigquery-etl
-          tag: ${CIRCLE_TAG:-latest}
+      - when:
+          condition: *deploy-condition
+          steps:
+            - gcp-gcr/push-image: *private-image
   manual-trigger-required-for-fork:
     docker: *docker
     steps:
@@ -496,11 +501,6 @@ workflows:
             # webhooks used in Ops logic. For details, see:
             # https://bugzilla.mozilla.org/show_bug.cgi?id=1715628#c0
             - deploy-to-private-gcr
-          filters:
-            branches:
-              only: main
-            tags:
-              only: /.*/
       # The following "private" jobs are basically clones of the public jobs
       # for generate-sql, deploy, and push-generated-sql, except that they pull
       # in some additional content from an internal Mozilla repository for
@@ -523,8 +523,3 @@ workflows:
             - private-generate-sql
             # can't run in parallel because CIRCLE_BUILD_NUM is same
             - build
-          filters:
-            branches:
-              only: main
-            tags:
-              only: /.*/

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,7 +1,15 @@
 ARG PYTHON_VERSION=3.8
 
+# use buster image because the default bullseye image released 2021-08-17
+# sha256:ffb6539b4b233743c62170989024c6f56dcefa69a83c4bd9710d4264b19a98c0
+# has updated coreutils that require a newer linux kernel than provided by CircleCI, per
+# https://forums.docker.com/t/multiple-projects-stopped-building-on-docker-hub-operation-not-permitted/92570/6
+# and https://forums.docker.com/t/multiple-projects-stopped-building-on-docker-hub-operation-not-permitted/92570/11
+FROM python:${PYTHON_VERSION}-slim-buster AS base
+WORKDIR /app
+
 # build typed-ast in separate stage because it requires gcc and libc-dev
-FROM python:${PYTHON_VERSION}-slim
+FROM base AS python-deps
 RUN apt-get update -qqy && apt-get install -qqy gcc libc-dev
 COPY java-requirements.txt ./
 RUN pip install -r java-requirements.txt
@@ -9,21 +17,19 @@ COPY requirements.txt ./
 RUN pip install -r requirements.txt
 
 # download java dependencies in separate stage because it requires maven
-FROM python:${PYTHON_VERSION}-slim
-# man is directory removed in upstream debian:buster-slim, but needed by jdk install
+FROM base AS java-deps
+# man directory is removed in upstream debian:buster-slim, but needed by jdk install
 RUN mkdir -p /usr/share/man/man1 && apt-get update -qqy && apt-get install -qqy maven
-WORKDIR /app
 COPY pom.xml ./
 RUN mvn dependency:copy-dependencies
 
-FROM python:${PYTHON_VERSION}-slim
+FROM base
 # add bash for entrypoint and jdk for jni access to zetasql
 RUN mkdir -p /usr/share/man/man1 && apt-get update -qqy && apt-get install -qqy bash default-jdk-headless
 COPY --from=google/cloud-sdk:alpine /google-cloud-sdk /google-cloud-sdk
 ENV PATH /google-cloud-sdk/bin:$PATH
-COPY --from=1 /app/target/dependency /app/target/dependency
-COPY --from=0 /usr/local /usr/local
-WORKDIR /app
+COPY --from=java-deps /app/target/dependency /app/target/dependency
+COPY --from=python-deps /usr/local /usr/local
 COPY .bigqueryrc /root/
 COPY . .
 RUN pip install .

